### ISSUE TITLE
Simplify Netty pipeline using Buffering handler

### ISSUE THREAD

**Issue by nmittler** [grpc/grpc-java#789]
The current pipeline configurations using one of the buffering handlers (tls, plaintext, etc.) are rather complicated and it's not always clear which handlers will handle exceptions in various cases.

We currently add the buffering handler and the HTTP/2 handler at startup. The buffering handler holds any writes until the startup handshake (e.g. SSL/TLS) completes, at which point it directs all buffered writes to the HTTP/2 handler. While those writes are occuring, the buffering handler stays in the pipeline (this is due to threading behavior of Netty WRT writes occuring outside of the event loop). If any problems occur while those writes are taking place, exception handling could occur in either the buffering handler or the HTTP/2 handler.  It would be desirable to guarantee that exception handling can occur in only a single place at any point in time.

Proposed change:

Part 1): Add a ChannelHandlerAdapter as the last handler in the pipeline. Netty has a race condition when writes occur from outside of the event loop.  The last ChannelHandlerContext is extracted in this thread and then the write is called. If however, the pipeline is changed between when the context is obtained and the write occurs ... badness ensues. As a workaround, there is some hacky code in the buffering handler to account for this race. A better solution to this problem would be to simply enforce the existence of a handler at the tail of the pipeline which never changes.  This will just be a pass-through, but must implement the `write` method (this is to avoid another Netty gotcha, where it will skip handlers if it has determined that they are uninterested in the event).

Part 2): With the handler from Part 1 in place, the installation of the buffering and HTTP/2 handlers can be modified to a `replace`.  Initially, only the buffering handler is installed (not the HTTP/2 handler). When the startup handshake completes successfully, the buffering handler will replace itself with the HTTP/2 handler, and then empty it's queued writes to the HTTP/2 handler.

In this way we guarantee that only one of these handlers exists in the pipeline at a time.  Failures due to the initial handshake will be handled by the buffering handler. Failures due to writes will always be handled by the HTTP/2 handler.

@ejona86 FYI

**Comment by buchgr** [comment]
> When the startup handshake completes successfully, the buffering handler will replace itself with the HTTP/2 handler, and then empty it's queued writes to the HTTP/2 handler.

@nmittler If I recall correctly I also tried to go along this path, but there was the issue that a write to the HTTP/2 handler can trigger other writes (I forgot why that actually happens, but it's easy to see that it's theoretically possible). So if the buffering handler is removed before the buffered write queue is empty, then those writes will mix with the still buffered writes and we lose order. Does that make sense?

**Comment by nmittler** [comment]
@buchgr, yeah Eric had mentioned that. It's not exactly clear to us how that might happen either. But assuming it does happen, I believe it should be OK so long as we replace the handler before doing any writes. Then send all of the writes to t he HTTP/2 handler.  Do you think that would work, or am I missing something?

**Comment by buchgr** [comment]
@nmittler replacing / removing the buffering handler before doing the writes is the problem. If the buffering handler remains there, then any new writes will be appended to the end of the buffering handler queue. If it is not there new writes will sneak in between, since everything is callbacks i.e. 

If your queue contains `bufferedWrite1, bufferedWrite2, bufferedWrite3`, and `bufferedWrite1` triggers two new writes then what you want is them to be written in the order `bufferedWrite1, bufferedWrite2, bufferedWrite3, newWrite1, newWrite2`, but what you get with removing the buffering handler before is

```
write(bufferedWrite1)
   ...
  write(newWrite1)
    ...
    write(newWrite2)
write(bufferedWrite2)
write(bufferedWrite3)
```

I think if such a mixup of writes happens in a real world application, it will be almost impossible to debug.

**Comment by nmittler** [comment]
@buchgr ah I see ... so the additional writes would happen immediately since we're in the event loop.  It's still strange as to why a write would trigger further writes.  Any ideas why this might happen?

**Comment by buchgr** [comment]
@nmittler I forgot sorry. I just remember that it did happen occasionally and that after figuring out why, it actually made sense :-).

**Comment by nmittler** [comment]
@buchgr Ha! Ok, might still be worth going through this to rediscover why it doesn't work ... and maybe adding some javadoc.

**Comment by buchgr** [comment]
@nmittler yeah I should have been more explicit in this comment https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java#L287 . Sorry :-)!

If I recall correctly you have a good chance of triggering this behavior by simply moving `ctx.pipeline().remove(this);` to before the writes and running the integration tests.
